daemon/containerd/image.go

```diff
@@ -248,37 +257,37 @@ func (i *ImageService) resolveDescriptor(ctx context.Context, refOrID string) (o
 		}
 		imgs, err := is.List(ctx, filters...)
 		if err != nil {
-			return ocispec.Descriptor{}, err
+			return containerdimages.Image{}, err
 		}
 
 		if len(imgs) == 0 {
-			return ocispec.Descriptor{}, images.ErrImageDoesNotExist{Ref: parsed}
+			return containerdimages.Image{}, images.ErrImageDoesNotExist{Ref: parsed}
 		}
 		if len(imgs) > 1 {
 			digests := map[digest.Digest]struct{}{}
 			for _, img := range imgs {
 				if img.Name == ref {
-					return img.Target, nil
+					return img, nil
 				}
 				digests[img.Target.Digest] = struct{}{}
 			}
 
 			if len(digests) > 1 {
-				return ocispec.Descriptor{}, errdefs.NotFound(errors.New("ambiguous reference"))
+				return containerdimages.Image{}, errdefs.NotFound(errors.New("ambiguous reference"))
 			}
 		}
 
-		return imgs[0].Target, nil
+		return imgs[0], nil
 	}
 
 	img, err := is.Get(ctx, ref)
```

Comment:
I think this one should be checked before `if truncatedID.MatchString(refOrID) {`, so that we always pick the exact-name match first.

Example:
```
$ docker images
REPOSITORY   TAG       IMAGE ID       CREATED       SIZE
random       latest    abcd0588ff2c   3 weeks ago   7.46MB
abcd         latest    3fbaf71a998b   4 weeks ago   3.73MB

$ docker rmi abcd
```


In this case, graphdriver Docker will delete only the `abcd` image:

```
$ docker rmi abcd
Untagged: abcd:latest
Deleted: sha256:3fbaf71a998bae6e375be74b999bd418091bf6511e356a129fdc969c4a94a5bc
Deleted: sha256:abcd052d13ad7f11c2b56249c179249f1ce37b0e1caf453bb037d863fd2588072

$ docker images
REPOSITORY   TAG       IMAGE ID       CREATED       SIZE
random       latest    abcd0588ff2c   3 weeks ago   7.46MB
```

Reply:
Good catch, yeah! I had another branch with a larger refactor that did that, but ended up trying to keep the exact same previous logic here for easier reviewing. I'm happy with that change though.

Reply:
✅